### PR TITLE
Split Router::extensions() combined setter/getter

### PR DESCRIPTION
**Pull request by dereuromark**
One of the last combined set/get methods in the core.
With this, we should have a clean API now, using the set* part for setting, and the normal method for getting in line with the others we split.

Also consistent with RouteBuilder usage

    $routes->scope('/', function (RouteBuilder $routes) {
        $routes->setExtensions(['json', 'xml']);
    });

### CODE REVIEW

**Comment by ADmad** [comment]
I would rather deprecate extensions setting using Router's static methods similar to the other methods and use Routebuilder instead. 

The only reason I haven't done so is because RequestHandlerComponent uses `Router::extensions()`. Now that I think about it, it can just use the extension of the route matched for the current request instead.

File: src/Routing/Router.php [hunk 1]
```diff
@@ -720,38 +720,68 @@ public static function normalize($url = '/'): string
...
+     *   Defaults to `true`. Deprecated since 4.3.0.
      * @return array<string> Array of extensions Router is configured to parse.
      */
     public static function extensions($extensions = null, $merge = true): array
     {
-        $collection = static::$_collection;
-        if ($extensions === null) {
-            return array_unique(array_merge(static::$_defaultExtensions, $collection->getExtensions()));
+        if ($extensions !== null) {
+            deprecationWarning('Setting extensions is deprecated, use setExtensions() instead.');
```
Comment (markstory) [inline src/Routing/Router.php]:
```suggestion
            deprecationWarning('Setting extensions with extensions() is deprecated, use setExtensions() instead.');
```

**Comment by ADmad** [comment]
> Now that I think about it, it can just use the extension of the route matched for the current request instead.

Actually `RequestHandlerComponent` already check's the currently matched route's extension but if not set it uses the list returned by `Router::extensions()` to decide whether it should set the extension based on the `Accept` header for matching type.

So we could add an option to `RequestHandlerComponent` for specifying the extensions/types it's interested in instead having it use `Router::extensions()`.

**Comment by ADmad** [comment]
TLDR; `Router::extensions()` as a getter is rarely required by users and it's use in `RequestHandler` can be removed, so the method split doesn't provide much value IMO. We either try to deprecated it altogether or keep it as it is and sort it out in `5.x`.

**Comment by markstory** [comment]
> Actually RequestHandlerComponent already check's the currently matched route's extension but if not set it uses the list returned by Router::extensions() to decide whether it should set the extension based on the Accept header for matching type.

I'm not sure this will be 100% backwards compatible. Right now RequestHandler will allow extension types on routes that don't have them enabled due to `Router::extensions()` being used.

Limiting extension detection to only the types that are enabled for the route seems like a good change to me. as we remove types bleeding into actions that shouldn't have extensions enabled, but it is something we would need to document.

**Comment by ADmad** [comment]
Yes, currently the types do unnecessarily bleed into all actions which might not actually support them. Not being able to maintain 100% backwards compatibility or any easy way to update existing code is why I chose to keep `Router::extensions()` and not depreciate like the other methods.

I think it would be best to add the new option to `RequestHandlerComponent` for limiting extensions/types in 5.0, where we don't have the burden of maintaining 100% BC.

So to reiterate, splitting out the setter into a separate method doesn't bring much value if setting/getting extensions globally through router is going to be dropped altogether in 5.0.

**Comment by markstory** [comment]
> So to reiterate, splitting out the setter into a separate method doesn't bring much value if setting/getting extensions globally through router is going to be dropped altogether in 5.0.

I agree. Moving towards the route controlling which extensions are valid is a good path forward.

**Comment by dereuromark** [comment]
I am fine with an alternative PR that deprecates it all, so you can remove in 5.x

**Comment by markstory** [comment]
Part of the replacement for this is in #15971

**Comment by markstory** [comment]
Based on the discussion in #15971 should we close this and revisit in 4.4 when we can come up with a replacement for RequestHandler and fully deprecate the `Router::extensions()` without a static method replacement?

**Comment by ADmad** [comment]
Replacing `RequestHandlerComponent` and not relying on globally set extensions will make `Router::extensions()` redundant. So let's keep it unchanged and deprecate it when we have an alternative to the component ready.

**Comment by othercorey** [comment]
@dereuromark Is it ok to close this do you want to still merge part of it?
